### DISCUSSION TITLE
Initialising Alpine components when using Livewire navigate

### DISCUSSION THREAD

**Discussion by jacksleight** [alpinejs/alpine#3880] [1. Help] [answered]
I have a component that uses a component object defined with `Alpine.data()`.

If I initialise that object inside an `alpine:init` event listener (as per https://alpinejs.dev/globals/alpine-data) and load the page directly everything works great.

However, if I navigate to the page using Livewire's navigate feature it is not initialised, as `alpine:init` doesn't fire after the first load. The Livewire docs have a section about not using `DOMContentLoaded` (https://livewire.laravel.com/docs/navigate#dont-rely-on-domcontentloaded), and I guess the same applies here.

So I thought the solution would be to listen to the `livewire:navigated` event instead, and that does make it work when I Livewire navigate to the page.

However, while `livewire:navigated` does fire when a page is loaded directly, it's fired after Alpine is started, which means the components try to initialise before `Alpine.data()` is called and I get "[component] not defined" errors.

Is there another way to do this that works on both direct page loads and Livewire navigate loads?

I would use Livewire's new `@script` directive, but it only works inside Livewire components and I'm trying to do this on regular pages.

**Comment by ekwoka** [comment, marked as answer]
Just put it in both.

Make one function that does it, and put it in both listeners.

**Reply by jacksleight** [reply]
Ah OK. That would mean it'll run twice on a direct page load, but I guess that doesn't really matter? The second `Alpine.data()` call will just override the first one.

**Reply by ekwoka** [reply]
Yeah, it won't cause any issues. It would similarly happen if you navigate back to the same page.

**Reply by TonyHuzz** [reply]
Hi @jacksleight  @ekwoka,  I'm encountering this issue too. I'm using Livewire v3 and Laravel 11
I have a `init.js `

```
export default (cb) => {
    document.addEventListener('alpine:init', cb);
    document.addEventListener('livewire:init', cb);
    document.addEventListener('livewire:navigated', cb);
};
```


And in my datepicker.js file
I'm doing like this

```
import init from "@alpine/init.js";
import flatpickr from "flatpickr";
import "flatpickr/dist/flatpickr.min.css";

(() => init(
    () => Alpine.data('datepicker', ({mode, showMonths, defaultDate}) => ({
        flatpickr: null,
        mode,
        showMonths,
        defaultDate,
        init() {
            ...
        },
    }))
))();

```

But still getting "[component] not defined" errors.
Am I doing wrong ?


Note: sry, I relized that it only occurs when I set Livewire navigate to true to let redirect looks like SPA.
But is there any solution for Alpine.data() initialize when using Livewire navigate redirects like SPA?

**Reply by jacksleight** [reply]
This is what I did in the end. I don't really like it, feels like there should be better way, but seems to work OK:

```js
document.addEventListener(typeof Alpine === 'undefined' ? 'alpine:init' : 'livewire:navigated', function() {
    // ...
}, { once: true });
```

It runs before any other scripts are loaded, which is why the `Alpine === 'undefined'` check works for the initial page load.

**Reply by TonyHuzz** [reply]
@jacksleight  Hi, but if u doing like this

`
$this->redirectIntended(default: route('dashboard', absolute: false), navigate: true);
`

will it be working too ?
my situation is that when I login to the page that has the component using above sentence, it will fail to init
I guess is because of this [Livewire redirect](https://livewire.laravel.com/docs/navigate#redirects)
but if I change to this

`$this->redirectIntended(default: route('dashboard', absolute: false), navigate: false);
`

It will init properly...

**Reply by jacksleight** [reply]
Yep, when I redirect as per your example the `livewire:navigated` event is fired.

**Reply by TonyHuzz** [reply]
Okay.. that's weird. cuz thougth my `livewire:navigated` event fired too, but I still getting "[component] not defined" errors.
I'm going to have a look again. Appreciate ur answer sir,  thanks a lot.

**Reply by mauritskorse** [reply]
@TonyHuzz: have you found a solution?
Ik keep having this issue as well after a fresh page load on another page, then navigating (using wire:navigate) to the page for the first time, the alpine component doesn't load. After visting the page for a second time without a page refresh the component works.

With the solution @jacksleight suggested I still keep getting this error, even though the eventlistener was called right before the error occurs:

```log
> 'livewire:navigated'
> livewire.js?id=38dc8241:1120 Alpine Expression Error: test is not defined
```` 
 This is my blade component:        
```blade
<div  x-data="test"></div>
@push('scripts')
<script>
document.addEventListener(typeof Alpine === 'undefined' ? 'alpine:init' : 'livewire:navigated', function() {
    console.log(typeof Alpine === 'undefined'? 'alpine init ' : 'livewire:navigated');
    Alpine.data('test', () => ({
        init() {
            console.log('initiating test');
        },
    }))
}, { once: true });
</script>
@endPush
```
I have tried with and without the push to the scripts stack and moving the scripts stack all the way to the <head> tags and towards the </body> tag. But nothing seems to change.

**Reply by TonyHuzz** [reply]
@mauritskorse Sry, I haven't found a properly solution. Maybe @jacksleight's solution is worked but my construct is wrong so not working for me.
My workaround is to avoid using wire:navigate for now, since it's not necessary.
